src/main/cpp/src/slice.cu

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "slice.hpp"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sizes_to_offsets_iterator.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/lists/filling.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/logical.h>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+namespace detail {
+
+namespace {
+
+void assert_start_is_not_zero(column_device_view const& start, rmm::cuda_stream_view stream)
+{
+  bool start_valid = thrust::all_of(rmm::exec_policy(stream),
+                                    thrust::make_counting_iterator(0),
+                                    thrust::make_counting_iterator(start.size()),
+                                    [start] __device__(size_type index) {
+                                      if (start.is_null(index)) return true;
+                                      return start.element<int32_t>(index) != 0;
+                                    });
+  CUDF_EXPECTS(start_valid, "Invalid start value: start must not be 0");
+}
+
+void assert_length_is_not_negative(column_device_view const& length, rmm::cuda_stream_view stream)
+{
+  bool length_valid = thrust::all_of(rmm::exec_policy(stream),
+                                     thrust::make_counting_iterator(0),
+                                     thrust::make_counting_iterator(length.size()),
+                                     [length] __device__(size_type index) {
+                                       if (length.is_null(index)) return true;
+                                       return length.element<int32_t>(index) >= 0;
+                                     });
+  CUDF_EXPECTS(length_valid, "Invalid length value: length must be >= 0");
+}
+
+struct int_iterator_from_scalar {
```

Comment:
The class name is a little confusable. It plays the role of an accessor more than an iterator.

`int_scalar_accessor`, and `int_column_accessor` would be better.